packages/x-data-grid-pro/src/hooks/features/dataSource/useGridDataSource.ts

```diff
@@ -85,39 +94,66 @@ export const useGridDataSource = (
         apiRef.current.resetDataSourceState();
       }
 
-      const fetchParams = gridGetRowsParamsSelector(apiRef);
-
+      const fetchParams = { ...gridGetRowsParamsSelector(apiRef), ...rowFetchSlice.current };
+      const startingIndex = fetchParams.start;
       const cachedData = apiRef.current.unstable_dataSource.cache.get(fetchParams);
 
       if (cachedData !== undefined) {
         const rows = cachedData.rows;
-        apiRef.current.setRows(rows);
-        if (cachedData.rowCount) {
+        if (cachedData.rowCount !== undefined) {
           apiRef.current.setRowCount(cachedData.rowCount);
         }
+        if (isLazyLoaded) {
+          apiRef.current.unstable_replaceRows(startingIndex, rows);
+        } else {
+          apiRef.current.setRows(rows);
+        }
+        apiRef.current.publishEvent('rowsFetched');
         return;
       }
 
+      // with lazy loading, only the initial load should show the loading overlay
+      const useLoadingIndicator = !isLazyLoaded || apiRef.current.getRowsCount() === 0;
       const isLoading = gridRowsLoadingSelector(apiRef);
-      if (!isLoading) {
+      if (!isLoading && useLoadingIndicator) {
         apiRef.current.setLoading(true);
       }
 
       try {
         const getRowsResponse = await getRows(fetchParams);
         apiRef.current.unstable_dataSource.cache.set(fetchParams, getRowsResponse);
-        if (getRowsResponse.rowCount) {
+        if (getRowsResponse.rowCount !== undefined) {
           apiRef.current.setRowCount(getRowsResponse.rowCount);
         }
-        apiRef.current.setRows(getRowsResponse.rows);
+        if (isLazyLoaded) {
+          apiRef.current.unstable_replaceRows(startingIndex, getRowsResponse.rows);
+        } else {
+          apiRef.current.setRows(getRowsResponse.rows);
+        }
         apiRef.current.setLoading(false);
+        apiRef.current.publishEvent('rowsFetched');
       } catch (error) {
         apiRef.current.setRows([]);
```

Comment:
Is there anything in place for the regular server side data?
I guess we can use the same on both, but leave the existing rows in place for lazy loading

Reply:
> Is there anything in place for the regular server-side data?

We do have [one](https://mui.com/x/react-data-grid/server-side-data/tree-data/#error-handling) for tree-data lazy loading, for the plain server-side data there isn't anyone because it always loads all the rows for the current page which could use a full-page [error overlay](https://mui.com/x/react-data-grid/server-side-data/#error-handling) (check `Make the requests fail` and re-fetch rows).

But for lazy loading and infinite loading, a full-page error overlay doesn't make much sense (not even for the first page) since it will block the users from seeing other loaded pages.

Coming up with a nice error UX for the plain lazy-loaded rows would be tricky. As we have to show the rows in place of the skeleton rows when the rows are available. I haven't seen any application with a perfect UX in this regard.

I can think of a few possibilities.

1. Do nothing, rely on the user scrolling out of the viewport and back on the same area of the viewport to retry fetching the rows.
2. Change some styling on the skeleton rows. Like the background color of the skeleton row to be red and adding a new row on top of the failed row section with a retry button (that will be auto removed when the user scrolls back to this section and a retry call has been sent.
3. Provide a demo with the `unstable_onDataSourceError` by implementing a persistent Snackbar approach that includes a Retry button (we'll possibly have to rethink the API then too, to allow the user to request a specific portion of the data, let's say rows from index 100 to 109)
4. We keep track of failed pages in the state, allow re-fetch using an API, and have a button on top of the demo saying `Refetch Failed Batches`

@KenanYusuf I'd be interested in what you think about this since you recently touched the skeleton rows part.

Reply:
Sorry for the late reply here, missed the notification.

> 3. Provide a demo with the unstable_onDataSourceError by implementing a persistent Snackbar approach that includes a Retry button (we'll possibly have to rethink the API then too, to allow the user to request a specific portion of the data, let's say rows from index 100 to 109)

I think this is a good option, it's non-disruptive, and allows users to carry on interacting with the data grid despite the fetching error. Updating the API to allow for this would allow users to customize the error experience too.

Reply:
I also agree, that number 3 is a good option. We don't have to change much and developers are free to choose how to deal with this. My plan is to add a demo for this case. For now, there is an update for the `fetchRows` api.
https://github.com/mui/mui-x/blob/50d9f0b992d1d0be617e54675fe63f0c4aaac544/packages/x-data-grid-pro/src/hooks/features/dataSource/interfaces.ts#L36
It changes the signature allowing more options to be passed. The data source is unstable, but I guess people started adopting it, so I was thinking throwing a warning first and remove the old signature when made stable
https://github.com/mui/mui-x/blob/50d9f0b992d1d0be617e54675fe63f0c4aaac544/packages/x-data-grid-pro/src/hooks/features/dataSource/useGridDataSource.ts#L122-L133